src/librustc_mir/transform/type_check.rs

```diff
@@ -1240,6 +1242,31 @@ impl<'a, 'gcx, 'tcx> TypeChecker<'a, 'gcx, 'tcx> {
         }
     }
 
+    fn prove_aggregate_predicates(
+        &mut self,
+        aggregate_kind: &AggregateKind<'tcx>,
+        location: Location,
+    ) {
+        let tcx = self.tcx();
+
+        let instantiated_predicates = match aggregate_kind {
+            AggregateKind::Adt(def, _, substs, _) => {
+                tcx.predicates_of(def.did).instantiate(tcx, substs)
+            }
+
+            AggregateKind::Closure(def_id, substs) |
+            AggregateKind::Generator(def_id, substs, _) => {
+                tcx.predicates_of(*def_id).instantiate(tcx, substs.substs)
```

Comment:
I realized that [this code](https://github.com/spastorino/rust/blob/017b3c5fe90b846da610b9372977517f8b65a275/src/librustc_mir/transform/type_check.rs#L1179-L1207) should probably move into this match arm